### ISSUE TITLE
[prefer-function-type][consistent-type-definitions] both rules together can lead to invalid code after fix

### ISSUE THREAD

**Issue by tao-cumplido** [typescript-eslint/typescript-eslint#3003]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-function-type": ["error"],
    "@typescript-eslint/consistent-type-definitions": ["error", "interface]
  }
}
```

```TS
type X = {
  (): void;
}
```

**Expected Result**

```ts
type X = () => void;
```

**Actual Result**
```ts
interface X () => void // Expression expected. ts(1109)
```

**Additional Info**

I suppose this might be difficult to fix, since both rules are in conflict. It feels like `prefer-function-type` should win in this case, but I guess that would require an awkward exception check in `consistent-type-definitions`. Maybe a note to the docs could be added instead.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.2` |
| `@typescript-eslint/parser`        | `4.14.2` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `14.15.4` |

**Comment by bradzacher** [comment]
This is a buggy interaction with ESLint if anything.
This will only occur if you do "autofix all rules" (eg `--fix` from the CLI).

Both rules provide fixers that make the smallest set of code changes possible:
`prefer-function-type`:
```
 type X = {
          ^
   (): void;
^^^  ^^    ^
 };
^^ 
```

`consistent-type-definitions`:
```
 type X = {
 ^^^^   ^^
   (): void;
 };
```

Note that the changes do not interact at all - whilst they are *very* close, they are not actually touching the same characters in the code.
So when ESLint applies all fixes at once - it at the fixes, decides that each individual fix does not actually conflict, so that means it's safe to make all the changes together.

This results in the broken code.

I would consider raising this as a bug with ESLint itself - it should be smarter about detecting fix ranges, and not applying fixes that are too close to one another within the same fix pass<sup>[1]</sup>.

---

[1] For reference, when you do `--fix`, ESLint actually does up to 10 fix passes on your code in an attempt to fix all rules as well as all rules that report due to fixes.
